src/urbit.ts

```diff
@@ -0,0 +1,468 @@
+  }
+
+  /**
+   * Constructs a new Urbit connection.
+   *
+   * @param url  The URL (with protocol and port) of the ship to be accessed
+   * @param code The access code for the ship at that address
+   */
+  constructor(url: string, code: string) {
+    this.uid = `${Math.floor(Date.now() / 1000)}-${Urbit.hexString(6)}`;
+    this.code = code;
+    this.url = url;
+    this.outstandingPokes = new Map();
+    this.outstandingSubscriptions = new Map();
+
+    return new Proxy(this, {
+      get(target: Urbit, property: string) {
+        // First check if this is a regular property
+        if (property in target) {
+          return (target as any)[property];
+        }
+
+        // Then check if it's a registered app
+        const app = Urbit.apps.get(uncamelize(property));
+        if (app) {
+          return new app(target);
+        }
+        
+        // Then check to see if we're trying to register an EventSource watcher
+        if (property.startsWith('on')) {
+          const on = uncamelize(property.replace('on', '')).toLowerCase();
+          return ((action: CustomEventHandler) => {
+            target.eventSource().addEventListener('message', (event: MessageEvent) => {
+              if (target.verbose) {
+                console.log(`Received SSE from ${on}: `, event);
+              }
+              if (event.data && JSON.parse(event.data)) {
+                const data: any = JSON.parse(event.data);
+                if (data.json.hasOwnProperty(on)) {
+                  action(data.json[on], data.json.response);
+                }
+              }
+            });
+          });
+        }
+
+        return undefined;
+      }
+    })
+  }
+
+  /**
+   * All-in-one hook-me-up.
+   * 
+   * Given a ship, url, and code, this returns an airlock connection
+   * that is ready to go. It `|hi`s itself to create the channel,
+   * then opens the channel via EventSource.
+   * 
+   * @param AuthenticationInterface
+   */
+  static async authenticate({ ship, url, code, verbose = false }: AuthenticationInterface) {
+    const airlock = new Urbit(`http://${url}`, code);
+    airlock.verbose = verbose;
+    airlock.ship = ship;
+    await airlock.connect();
+    await airlock.poke(ship, 'hood', 'helm-hi', 'opening airlock');
+    await airlock.eventSource();
+    return airlock;
+  }
+
+  /**
+   * Connects to the Urbit ship. Nothing can be done until this is called.
+   * That's why we roll it into this.authenticate
+   */
+  async connect(): Promise<void> {
+    if (this.verbose) {
+      console.log(`password=${this.code} `, isBrowser ? "Connecting in browser context at " + `${this.url}/~/login` : "Connecting from node context");
+    }
+    return fetch(`${this.url}/~/login`, {
+      method: 'post',
+      body: `password=${this.code}`,
+      credentials: 'include',
+    }).then(response => {
+      if (this.verbose) {
+        console.log('Received authentication response', response);
```

Comment:
An actual logging library of sorts might be a nice addition. For example, `hoon-language-server` needs to use the stdio to give/receive langserver stuff and `console.log` will get in the way there

Reply:
I'm open to suggestions